### DISCUSSION TITLE
Unable to invoke federated dynamic import a second time when first attempt fails

### DISCUSSION THREAD

**Discussion by ajelcocks** [webpack/webpack#16278] [Q&A]
Hello,
If I have a remote (federated) module embedded in my main application that fails when the application is started I would like to add a refresh button for just that module. However, the failed load is cached and returned. The only option is to refresh the browser.

There is an example here ... https://github.com/ajelcocks/federated-retry

Does anyone know if there any way to force a reload?
or 'flush' the item from the (internal) __webpack_module_cache__, for example

Or should I add this as an issue and suggest the code be changed to retry if the cached item is in error

**Comment by alexander-akait** [comment]
You can reload browser, so just catch the error and reload the problem

**Reply by ajelcocks** [reply]
Thankyou for the reply.
I know I can refresh the browser. However, that will refresh everything and lose state, or require re-establishment of state
Suppose I have 5 areas on the page, each supplied by a different micro-frontend.
One is temporarily unavailable.
I want to be able to refresh the one item rather than the whole page.
I could do that with iframes, but that presents its own challenges.
I am specifically trying to do this using federated modules

I am researching a way to be able to use the dynamic import ... import(url).then(() => { do stuff }
but once that fails I have no way of flushing that from the webpack cache

The options I see would be a way for my code to remove that item from the webpack cache, which is not exposed, or for the webpack code to not cache a failed fetch

**Reply by alexander-akait** [reply]
I understand your problem, but it lies far beyond webpack, we don't know the structure of your application, how you load it and its states, so unfortunately I can't give an answer here, you can always rewrite webpack runtime code and implement own logic for it, please read it here https://webpack.js.org/api/module-variables/

**Reply by ajelcocks** [reply]
Regarding my code and its structure, I took the time to write the simplest example code I could and post it. See the link in the initial reply.

Furthermore, regarding it being far beyond webpack, I found the line(s) of code responsible at https://github.com/webpack/webpack, assuming that is the webpack code.

When the dynamic import statement is invoked the webpack code I found creates an entry in __webpack_module_cache__ even when the http request fails.
Subsequent invocations then just return the error, rather than retry the request

I believe the code in question is generated by https://github.com/webpack/webpack/blob/main/lib/javascript/JavascriptModulesPlugin.js near line 1040

The part I think is the culprit is...

function __webpack_require__(moduleId) {
	// Check if module is in cache
	var cachedModule = __webpack_module_cache__[moduleId];
	if (cachedModule !== undefined) {
		if (cachedModule.error !== undefined) throw cachedModule.error;
		return cachedModule.exports;
	}


Is this not webpack code? ... and it detects the last import was an error and returns it again? .. or am I missing something?

**Reply by alexander-akait** [reply]
You can override `webpack_require`, and you can delete manually module from `webpack_module_cache`

**Reply by ajelcocks** [reply]
I'm not sure how to override webpack_require.
The function in question is __webpack_require__(moduleId)
This is contained in code generated by webpack as a self executing function
(() => { // webpackBootstrap
})()
Thus the vars and functions declared within are scoped to the IIEF
I do not see where that function is exposed
Could you explain how I override webpack_require?
I have looked in both the window and document object and see nothing that looks relevant

I could fork webpack, as you suggested, but I would not think that is recommended. I certainly don't want to start maintaining and managing a fork

My need is simple. I would like to be able to invoke dynamic import and have it retry the fetch if the last fetch yielded an error.
The webpack code currently prevents that, simply returning the error again.

My issue would be resolved if webpack exposed an api at runtime facilitating removing the error from the webpack_module_cache, or a flag forcing reload in the dynamic import, for instance import("foo", true)

Thanks

**Reply by alexander-akait** [reply]
Do you need this - https://webpack.js.org/api/module-variables/#__webpack_chunk_load__-webpack-specific, we have an example...
